openapi.yaml

```diff
@@ -0,0 +1,2201 @@
+        '400':
+          $ref: '#/components/responses/BadRequest'
+        '401':
+          $ref: '#/components/responses/Unauthenticated'
+        '403':
+          $ref: '#/components/responses/PermissionDenied'
+
+  /pools/{pool_name}:
+    parameters:
+      - $ref: '#/components/parameters/PoolName'
+
+    get:
+      summary: Get a pool
+      operationId: getPool
+      tags: [Pool]
+      responses:
+        '200':
+          description: Successful response.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Pool'
+        '401':
+          $ref: '#/components/responses/Unauthenticated'
+        '403':
+          $ref: '#/components/responses/PermissionDenied'
+        '404':
+          $ref: '#/components/responses/NotFound'
+
+    patch:
+      summary: Update a pool
+      operationId: updatePool
+      tags: [Pool]
+      parameters:
+        - $ref: '#/components/parameters/UpdateMask'
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/Pool'
+
+      responses:
+        '200':
+          description: Successful response.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Connection'
+        '400':
+          $ref: '#/components/responses/BadRequest'
+        '401':
+          $ref: '#/components/responses/Unauthenticated'
+        '403':
+          $ref: '#/components/responses/PermissionDenied'
+        '404':
+          $ref: '#/components/responses/NotFound'
+
+    delete:
+      summary: Delete a pool
+      operationId: deletePool
+      tags: [Pool]
+      responses:
+        '204':
+          description: No content.
+        '400':
+          $ref: '#/components/responses/BadRequest'
+        '401':
+          $ref: '#/components/responses/Unauthenticated'
+        '403':
+          $ref: '#/components/responses/PermissionDenied'
+
+  /dags/{dag_id}/taskInstances:
+    parameters:
+      - $ref: '#/components/parameters/DAGID'
+      - $ref: '#/components/parameters/FilterExecutionDateStartRange'
+      - $ref: '#/components/parameters/FilterExecutionDateEndRange'
+      - $ref: '#/components/parameters/FilterStartDateStartRange'
+      - $ref: '#/components/parameters/FilterStartDateEndRange'
+      - $ref: '#/components/parameters/FilterEndDateStartRaange'
+      - $ref: '#/components/parameters/FilterEndDateEndRange'
+      - $ref: '#/components/parameters/FilterDurationStartRange'
+      - $ref: '#/components/parameters/FilterDurationEndRange'
+      - $ref: '#/components/parameters/FilterState'
+      - $ref: '#/components/parameters/FilterPool'
+      - $ref: '#/components/parameters/FilterQueue'
+    get:
+      summary: Get list of task instance of DAG.
+      description: >
+        This endpoint support reading resources across multiple DAGs by specifying a "-" as a dag_id.
```

Comment:
I tried to rely on Google's recommendations if I had doubts. In this case it is [AIP159](https://aip.dev/159).
> APIs may support reading resources across multiple collections by allowing users to specify a - (the hyphen or dash character) as a wildcard character in a standard List method:

This is still a topic of discussion and I will want to raise it on the mailing list.

Reply:
This won't work right now as `-` is allowed character in dag IDs, and I'd rather not have to change it just to support, what is (still to me) a very odd way of abusing a single param to fit in multiple values.

Reply:
Do we have a current use case where we want to read TIs across multiple dags, or is this a new use case?

Reply:
What do you think about `*` as a wildcard character? 

I would like the hierarchical structure to be preserved, because it can facilitate the use of this API.
We can change this endpoint
/dags/{dag_id}/taskInstances:
/dags/{dag_id}/taskInstances/{task_id}/{execution_date}:
Into:
/taskInstances/?dag_id={dag_id}:
/taskInstances/{task_id}/{execution_date}:
But then we will lose our url hierarchies completely and people who do not know Airflow will feel less confident.

Reply:
(sorry, on my phone so a bit difficult to look at much at once)

What about `/taskInstances/` that does anything across more than a single DagRun?

Reply:
It just feels like since TI is only "under" a dag (run) when it's a single dag, that when searching across multiple dags or runs, that is a search argument to a top level `/Ti` collection

Reply:
We have use case in current web UI:
http://0.0.0.0:28080/taskinstance/list/
I also have a user case in my project that we are currently developing. We run many DAGs that are not related to each other and after DAGs we want to know what tasks were successful and which were unsuccessful. 
Monitoring is another example. You can fetch all tasks from the date range, e.g. from yesterday, to calculate the average duration of tasks or other metrics. You are not interested in specific DAGs, but you are interested in tasks.

Reply:
That as tone described it to sounds like a top-level TI collection makes more sense. To me anyway

Reply:
This is probably one for the mailing list, not just you and I to decide :)

Reply:
It is worth looking at all endpoints to see what your proposal affects the API.
We have the following endpoints related to /dags/:
```
/dags
/dags/{dag_id}
/dags/{dag_id}/clearTaskInstances
/dags/{dag_id}/dagRuns
/dags/{dag_id}/dagRuns/{execution_date}
/dags/{dag_id}/structure
/dags/{dag_id}/taskInstances
/dags/{dag_id}/taskInstances/{task_id}/{execution_date}
/dags/{dag_id}/taskInstances/{task_id}/{execution_date}/links
/dags/{dag_id}/taskInstances/{task_id}/{execution_date}/logs/{task_try_number}
/dags/{dag_id}/taskInstances/{task_id}/{execution_date}/xcomEntries
/dags/{dag_id}/taskInstances/{task_id}/{execution_date}/xcomEntries/{key}
/dags/{dag_id}/tasks
/dags/{dag_id}/tasks/{task_id}
```
If we move taskInstance a level up we get the following structure.
```
/dags
/dags/{dag_id}
/dags/{dag_id}/clearTaskInstances
/dagRuns/
/dagRuns/{dag_id}/{execution_date}
/dags/{dag_id}/structure
/taskInstances
/taskInstances/{dag_id}/{task_id}/{execution_date}
/taskInstances/{dag_id}/{task_id}/{execution_date}/links
/taskInstances/{dag_id}/{task_id}/{execution_date}/logs/{task_try_number}
/xcomEntries
/xcomEntries/{dag_id}/{task_id}/{execution_date}/{key}
/dags/{dag_id}/tasks
/dags/{dag_id}/tasks/{task_id}
```
It seems to me that my proposal is much more readable and a person who does not know Airflow will not be able to use these API more easily.

Reply:
Sorry I wasn't clear - I was suggesting having TI collection in two places - top level and under dag (run)

Reply:
Let me craft a full resonse when I get to my computer this evening

Reply:
Then we will have two endpoint that will perform the same operations. This means that the URLs will not be ambiguous. And the end-user will have to decide which method is right for him. In practice, users will only use this endpoint, which will provide more opportunities.

Reply:
I agree. I want to write about it on the list and I will do it today. I will also prepare a changelog with changes.